### PR TITLE
Custom data model fix

### PR DESCRIPTION
**Pull request by BrennenMM7**
**What type of PR is this?**

**What this PR does / why we need it**:
When Bootstrap token refreshes the VMSS needs to have the latest model in order for autoscaling machinepools to take place succesfully. Stores a hash of the current custom data contents into AzureMachinePool.Status.CustomDataHash. This allows to update the VMSS data if the custom data changes.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2803 

**Special notes for your reviewer**:
This is utilizing the same logic as created in this pull request [https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2803](url), the only changes are as follows:
- Added to the mock tests to past all unit testing, 
- Removed logic from node surge to prevent constant node rolling updates. 
- Added the watcher to the Kubeadmconfig resource as mentioned in the issue/PR linked. 
- API resource conversions for v1alpha3 and v1alpha4.
_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [X] squashed commits
- [ ] includes documentation
- [X] adds unit tests

**Release note**:

```release-note
- AzureMachinePool Controller will no watch KubeadmConfig to ensure AzureMachinePool Bootstrap data is updated on the VMSS
- The patch loop for the AzureMachinePool will now compare CustomData field to ensure the VMSS instance is updated with the latest model 
```

### CODE REVIEW

**Comment by CecileRobertMichon** [comment]
Should we consider only doing this when the MachinePool has the externally managed replicas annotation? I'd be worried about increasing the number of API calls for users that don't need this as this will now make a PUT API call per AzureMachinePool every 10 minutes or so (or whenever token is refreshed, I forget the exact number) which might not scale well for very large clusters with hundreds of MachinePools (granted this will be a problem for users with externally managed replicas too but if we can at least reduce the blast radius that would be an improvement).

**Comment by CecileRobertMichon** [comment]
@BrennenMM7 build is failing, PTAL

**Comment by BrennenMM7** [comment]
@CecileRobertMichon Fixed the build failures and squashed commits again, should be good to go. 👍

File: exp/api/v1beta1/azuremachinepool_types.go
Comment (CecileRobertMichon) [inline exp/api/v1beta1/azuremachinepool_types.go:313]: Would an annotation be a better fit for this?

File: azure/services/scalesets/scalesets.go
Comment (CecileRobertMichon) [inline azure/services/scalesets/scalesets.go:315]: this could be 

```
if *patch.Sku.Capacity <= infraVMSS.Capacity && !hasModelChanges && (!hasCustomDataChanges || !replicaseManagedByAutoscaler)  {
```

to avoid the duplication

File: azure/services/scalesets/scalesets.go
Comment (CecileRobertMichon) [inline azure/services/scalesets/scalesets.go:317]:
```suggestion
			log.V(4).Info("nothing to update on vmss", "scale set", spec.Name, "newReplicas", *patch.Sku.Capacity, "oldReplicas", infraVMSS.Capacity, "hasModelChanges", hasModelChanges, "hasCustomDataChanges", hasCustomDataChanges)
```

File: azure/services/scalesets/scalesets.go
Comment (CecileRobertMichon) [inline azure/services/scalesets/scalesets.go:293]: Have you checked if we get customData back from the existing VMSS? if so we might not even need to store the custom data hash since we would be able to just compare the custom data we have with the existing VMSS custom data

File: exp/api/v1beta1/azuremachinepool_types.go
Comment (CecileRobertMichon) [reply exp/api/v1beta1/azuremachinepool_types.go:313]: this might be superseded by https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3134#discussion_r1107459190

File: azure/services/scalesets/scalesets.go
Comment (BrennenMM7) [reply azure/services/scalesets/scalesets.go:293]: We do return the customData from the fetchedVMSS state under the getVirtualMachineScaleSet() function, however the SDKtoVMSS converters remove the OSProfile data in which its stored. If this is the preferred method I can update the converters and types to facilitate this comparison rather than storing a hash.

Comment (CecileRobertMichon) [reply azure/services/scalesets/scalesets.go:293]: yes I think that would be a cleaner implementation and it can rely on the actual state of the world instead needing to store the hash in the API. We can modify the `SDKtoVMSS` converters helper function easily to do what we need since it's part of the CAPZ code.

File: exp/api/v1beta1/azuremachinepool_types.go
Comment (BrennenMM7) [reply exp/api/v1beta1/azuremachinepool_types.go]: Resolved by using the SDKtoVMSS vs the API

File: azure/services/scalesets/scalesets.go
Comment (BrennenMM7) [reply azure/services/scalesets/scalesets.go]: Consolidated if block

File: azure/services/scalesets/scalesets.go
Comment (BrennenMM7) [reply azure/services/scalesets/scalesets.go:280]: I'm not sure how else we would be able to differentiate the reconcile loop based on the existence of the external autoscaler annotation. Open to suggestions.

File: azure/services/scalesets/scalesets.go
Comment (BrennenMM7) [reply azure/services/scalesets/scalesets.go:284]: Ok I think I've gone through and solidified the logic to use the SDKtoVMSS data, just waiting for the tests to green light and I'll mark this as resolved.

File: azure/services/scalesets/scalesets.go
Comment (CecileRobertMichon) [reply azure/services/scalesets/scalesets.go:280]: I just meant using `s.Scope.HasReplicasExternallyManaged(ctx)` directly instead of assigning it to a variable since it's only used once... not a huge deal though, definitely a nit

**Review by CecileRobertMichon — COMMENTED** [review]
This looks good @BrennenMM7 thanks for reworking it to remove the need to change the API

/lgtm
/assign @mboersma

**Comment by CecileRobertMichon** [comment]
/hold

Please add a release note in the PR description, this is definitely a user-facing change worth calling out

**Comment by BrennenMM7** [comment]
Release note added.

**Review by mboersma — APPROVED** [review]
/lgtm
/approve
/cherry-pick release-1.7

**Comment by BrennenMM7** [comment]
Actually, can we hold the PR before merging. I spoke with @dthorsen this morning and discovered that the API for VMSS in the CustomData field may always return a null field for security purposes and therefore would cause the machine pool to be patched every reconcile loop. 

The stored hash may be required as I suspected before. Need to verify.

**Comment by BrennenMM7** [comment]
@CecileRobertMichon Just confirmed that the AzureSDK will always return a nil value for "Properties.VirtualMachineProfile.OSProfile.CustomData" even if its not true to protect the data stored within that var on Get requests. In order to actually have a working reconcile loop compare the data, the old method of storing a hash into the API before calling the patch is the only way to compare. 

Unless we have an alternative method of providing this token data to the VMSS model for consumption or Microsoft changes the rules on that portion of the API response lol. 

Let me know and I can roll back to the previous commits. 👍

**Comment by dthorsen** [comment]
This is probably worth a discussion during the next office hours. There are some options to consider regarding how best to proceed with regard to API status fields, as well as with regard to documentation and some implications of this behavior.

**Comment by CecileRobertMichon** [comment]
@dthorsen @BrennenMM7 great catch... Apologies for leading you on this incorrect path. I'm still not a fan of adding this into the API since it's not meant to be consumed by the user, maybe a good compromise is storing the hash in an annotation?

**Comment by BrennenMM7** [comment]
@CecileRobertMichon Okay the logic has been reverted and is now stored inside an Annotation on the AMP object instead.

**Comment by CecileRobertMichon** [comment]
build is failing PTAL @BrennenMM7

**Comment by CecileRobertMichon** [comment]
looks like prow job was interrupted?

/retest

File: azure/scope/machinepool.go
Comment (CecileRobertMichon) [inline azure/scope/machinepool.go:542]:
```suggestion
			log.V(4).Error(err, "unable to update custom data hash, ignoring")
```

File: azure/scope/machinepool.go
Comment (CecileRobertMichon) [inline azure/scope/machinepool.go:607]: let's make the annotation name a const

File: azure/services/scalesets/scalesets.go
Comment (CecileRobertMichon) [inline azure/services/scalesets/scalesets.go:281]:
```suggestion
	replicasManagedByAutoscaler := s.Scope.HasReplicasExternallyManaged(ctx)
```

File: azure/services/scalesets/scalesets.go
Comment (CecileRobertMichon) [inline azure/services/scalesets/scalesets.go:285]: This error seems important, at least when `replicasManagedByAutoscaler` is true

Maybe another way to approach this:

```
hasCustomDataChanges := false
if s.Scope.HasReplicasExternallyManaged(ctx) {
  hasCustomDataChanges, err := s.Scope.HasBootstrapDataChanges(ctx) { 
    return nil, errors.Wrap(err, "unable to calculate custom data hash")
  }
  if hasCustomDataChanges {
    log.V(4).Info("custom data changed")
  } else {
    log.V(4).Info("custom data unchanged")
  }
}

File: azure/scope/machinepool.go
Comment (CecileRobertMichon) [reply azure/scope/machinepool.go:542]: > // ignore errors to calculating the custom data hash since it's not absolutely crucial.

not sure I agree, we should return all errors as much as possible, so the controller can requeue a new reconcile loop and retry 

this could lead to critical silent failure if the update is failing and causing a user to not be able to scale up a machine pool, or on the other hand throttling if the VMSS is trying to update on every reconcile because its hash annotation is out of date.

If you are concerned about impacting users who don't need this hash (when HasReplicasExternallyManaged is false), then maybe we can move this update inside an if statement to only make the update when it is needed (`if m.HasReplicasExternallyManaged(ctx)`) - this would make sense only if you apply this suggestion https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3134/files#r1114879941

File: azure/services/scalesets/scalesets.go
Comment (BrennenMM7) [reply azure/services/scalesets/scalesets.go]: Added change

File: azure/services/scalesets/scalesets.go
Comment (BrennenMM7) [reply azure/services/scalesets/scalesets.go]: Added change

File: azure/scope/machinepool.go
Comment (BrennenMM7) [reply azure/scope/machinepool.go]: Added const

File: azure/scope/machinepool.go
Comment (BrennenMM7) [reply azure/scope/machinepool.go]: Added if block to add a non impact to users not utilizing this feature.

File: azure/scope/machinepool.go
Comment (CecileRobertMichon) [inline azure/scope/machinepool.go:58]: let's move this to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/const.go where the other annotations are and use the same format and prefix, with a code comment that explains how it's used

```
sigs.k8s.io/cluster-api-provider-azure-vmss-custom-data-hash
```

**Review by CecileRobertMichon — COMMENTED** [review]
lgtm aside from comment on annotation const and failing unit test job, thanks again and sorry for all the back and forth on this one

**Comment by CecileRobertMichon** [comment]
/hold cancel

please squash!

**Comment by BrennenMM7** [comment]
Can I get a tag on this PR please 😄

**Review by nawazkh — COMMENTED** [review]
I don't fully understand the problem, will go back try reproducing the error. 
However, I had some doubts and questions jotted down below after reviewing the code.

File: azure/scope/machinepool.go
Comment (CecileRobertMichon) [reply azure/scope/machinepool.go:608]: the hash was moved to an annotation, see https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3134#issuecomment-1435315588

File: azure/services/scalesets/scalesets.go
Comment (CecileRobertMichon) [reply azure/services/scalesets/scalesets.go:291]: custom data changes will likely contain sensitive info such as the bootstrap token (which is why it's held in a secret) it should not be printed in logs

File: exp/controllers/helpers.go
Comment (CecileRobertMichon) [reply exp/controllers/helpers.go:348]: `Bootstrap` is not a pointer so it can never be nil (ref https://github.com/kubernetes-sigs/cluster-api/blob/0d8d072c045572504f4eca32f9fa9cd63d4289ee/api/v1beta1/machine_types.go#L65)

**Comment by CecileRobertMichon** [comment]
/lgtm
/hold

@nawazkh I think I've addressed all your questions/comments, let me know if you're good to remove the hold

**Comment by nawazkh** [comment]
Thanks for the updates @CecileRobertMichon ! 
Kudos @BrennenMM7 🎉 
/hold-cancel
/lgtm

**Comment by jackfrancis** [comment]
@BrennenMM7 did this PR fix #2683 as well?

Thank you!

cc @primeroz

**Comment by BrennenMM7** [comment]
@jackfrancis Yes this should resolve #2683 as the watcher will ensure that the VMSS always has the latest bootstrap token inside the model of the VMSS. 👍

**Comment by jackfrancis** [comment]
@BrennenMM7 thank you! I'll close #2683 and advise that the original reporter of that issue validate the fix with v1.8.0 after it is released (next week), because this PR will be included in that release.

Thanks again for the fix!

**Pull request merged**
